control/Session.php

```diff
@@ -595,6 +595,9 @@ public static function request_contains_session_id() {
 	 * @param string $sid Start the session with a specific ID
 	 */
 	public static function start($sid = null) {
+		// When sessions start, we assume that the responses should not be cached by CDNs / proxies as we are
+		// likely to be supplying information relevant to the current user only
+		HTTPCacheControl::singleton()->privateCache();
```

Comment:
This isn't going to capture all session starts; Only a small number (namely, those sessions started by errorcontrolchain during a flush).

Did you mean to put this into `Director::direct` where the inst_start() method is called?

```php
// Only resume a session if its not started already, and a session identifier exists
if(!isset($_SESSION) && Session::request_contains_session_id()) {
	$session->inst_start();
}
```

Also keep in mind that session can ALSO be started by the below code, which happens near the end of the request lifecycle (and possibly after cache-control has already been added to the response!)

```php
// Save session data. Note that inst_save() will start/resume the session if required.
$session->inst_save();
```

Instead of this approach, may I suggest another: In Controller::handleRequest we invoke this code.

```php
HTTP::add_cache_headers($response);
```

Should we instead check if `Session::getAll()` in `HTTP::add_cache_headers` to anticipate either:
 - A session has been started
 - A session will be started

And set privateCache() there instead?

Reply:
Your last suggestion sounds good.

Reply:
Note, you could also shift HTTP::add_cache_headers() into a postRequest filter as well, if you want to capture it as late as possible.